File: Recursion/Lecture 04/CodeStudio/Find power of a number.java
class Solution {

	public static long Pow(int X, int N) {
	// base condition
	if(N==0) return 1;

        long temp=Pow(X,N/2); // recursion call

        // if N is odd then we will have to multiply X 
        if(N%2==1) return temp*temp*X;
        return temp*temp;
	}
}


/*
    time complexity : O(log n)
    space complexity : O(log n)
*/

File: Recursion/Lecture 04/CodeStudio/Find power of a number.py
def pow(X, N):

    # base case
    if N == 0:
        return 1
    
    # recurssion call
    temp = pow(X, N//2) 
    
    # if N is odd then we will have to multiply x
    if N%2 == 1:
        return temp*temp*X
    
    return temp*temp

"""
Algorithm 

time complexity : O(log n)
space complexity : O(log n)
"""
